### DISCUSSION TITLE
How to limit number of records for preload() ?

### DISCUSSION THREAD

**Discussion by bartclaeys** [filamentphp/filament#4850] [Help]
How can I limit preload() to eg. the last 5 items instead of loading everything?

In below examples, I want to show the 5 recently added authors when creating a new book entry.

```
BelongsToSelect::make('books')
    ->relationship('author', 'name')
    ->preload()
    ->searchable()
```

**Comment by danharrin** [comment]
```php
Select::make('books')
    ->relationship('author', 'name', fn (Builder $query) => $query->latest())
    ->optionsLimit(5)
    ->preload()
    ->searchable()
```

**Comment by bartclaeys** [comment]
This still loads ALL records (but only showing 5). I want to show the latest 5 records, then use AJAX request to search the rest.

**Reply by danharrin** [reply]
Right.

```php
Select::make('books')
    ->relationship('author', 'name')
    ->options(fn (): array => Author::latest()->limit(5)->pluck('name', 'id')->all())
    ->searchable()
```

**Comment by bartclaeys** [comment]
Thanks! Almost there :) Now how do I combine this with `getOptionLabelFromRecordUsing()`?

**Reply by danharrin** [reply]
```php
Select::make('books')
    ->relationship('author', 'name')
    ->options(fn (Select $component): array => Author::latest()
        ->limit(5)
        ->get()
        ->map(fn (Author $record): string => $component->getOptionLabelFromRecord($record))
        ->all())
    ->getOptionLabelFromRecordUsing(fn (Author $record): string => "{$record->name} - {$record->location}")
    ->searchable()
```

**Comment by bartclaeys** [comment]
Unfortunately above code results in "The selected {model} is not valid" upon saving. I think it's missing an ID somewhere.

**Reply by jadersbr** [reply]
Did you find a solution?

**Reply by bartclaeys** [reply]
No, abandoned this for now, but will need to look into it again soon.

**Reply by oyenet1** [reply]
use ->toArray() method after the pluck

**Comment by mybouhssina** [comment]
It seems that preload is still loading all the relationship models (I tried with over 20k record) and it caused the loading to become extremely slow, desptie having to optionsLimit. I ended up setting the limit using modifyQueryUsing as follows:
...->relationship('categories', 'name', fn($query) => $query->limit(10))
and it seems to have solved the problem. I think the OptionsLimit is quite misleading.

**Reply by danharrin** [reply]
If its a bug, open an issue with a reproduction repository and we will investigate

**Reply by synsyst** [reply]
Wouldn't the $query->limit(10) cut off 11+ results?
